File: frontend/src/api/aClassAchievements.ts
import { api } from "./client";
import { appendAuthToken } from "../utils/fileLinks";

const A_CLASS_TABLE_ID = "tbl5hk9UpmisKOxC";

export type AClassKind = "pdf" | "image" | "extra";

export type AClassAttachment = {
  file_token: string;
  name?: string;
  size?: number;
  type?: string;
};

export type AClassAchievementRead = {
  id: number;
  base_record_id: string;
  project_content?: string;
  event_name?: string;
  level?: string;
  award_grade?: string;
  organizer?: string;
  event_date?: string;
  department?: string;
  responsible_person?: string;
  first_student?: string;
  other_students?: string;
  research_category?: string;
  pdf_files: AClassAttachment[];
  image_files: AClassAttachment[];
  extra_files: AClassAttachment[];
  ai_image_understanding?: string;
  kimi_summary?: string;
  notes?: string;
  created_at: string;
  updated_at: string;
};

export type ListResp = {
  items: AClassAchievementRead[];
  total: number;
  research_categories: string[];
  levels: string[];
};

export const listAClassAchievements = async (params?: {
  research_category?: string;
  level?: string;
}): Promise<ListResp> => {
  const { data } = await api.get<ListResp>("/a-class-achievements", { params });
  return data;
};

export const getAClassAchievement = async (id: number | string): Promise<AClassAchievementRead> => {
  const { data } = await api.get<AClassAchievementRead>(`/a-class-achievements/${id}`);
  return data;
};

export const uploadAClassFile = async (file: File): Promise<{
  file_token: string;
  name: string;
  size: number;
  type: string;
}> => {
  const form = new FormData();
  form.append("file", file);
  const { data } = await api.post("/files/upload?target=a_class", form, {
    headers: {
      "Content-Type": "multipart/form-data",
    },
  });
  return data;
};

export const patchAClassAttachments = async (
  baseRecordId: string,
  kind: AClassKind,
  fileTokens: string[],
): Promise<AClassAchievementRead> => {
  const { data } = await api.patch<AClassAchievementRead>(
    `/a-class-achievements/by-record/${encodeURIComponent(baseRecordId)}/attachments`,
    { kind, file_tokens: fileTokens },
  );
  return data;
};

export const aClassFileUrl = (fileToken: string, token?: string): string => {
  return appendAuthToken(`/api/files/${encodeURIComponent(fileToken)}/proxy?table_id=${encodeURIComponent(A_CLASS_TABLE_ID)}`, token);
};

File: backend/app/schemas/a_class_achievements.py
from __future__ import annotations

import json
from typing import Any

from pydantic import BaseModel, Field, field_validator


class AClassAttachment(BaseModel):
    file_token: str
    name: str | None = None
    size: int | None = None
    type: str | None = None


def _parse_attachments(raw: str | None) -> list[AClassAttachment]:
    if not raw:
        return []
    try:
        data = json.loads(raw)
    except (TypeError, ValueError):
        return []
    if not isinstance(data, list):
        return []
    out = []
    for x in data:
        if isinstance(x, dict) and x.get("file_token"):
            out.append(AClassAttachment(**{k: x.get(k) for k in ("file_token", "name", "size", "type")}))
    return out


class AClassAchievementRead(BaseModel):
    id: int
    base_record_id: str
    project_content: str | None = None
    event_name: str | None = None
    level: str | None = None
    award_grade: str | None = None
    organizer: str | None = None
    event_date: str | None = None
    department: str | None = None
    responsible_person: str | None = None
    first_student: str | None = None
    other_students: str | None = None
    research_category: str | None = None
    pdf_files: list[AClassAttachment] = Field(default_factory=list)
    image_files: list[AClassAttachment] = Field(default_factory=list)
    extra_files: list[AClassAttachment] = Field(default_factory=list)
    ai_image_understanding: str | None = None
    kimi_summary: str | None = None
    notes: str | None = None

    @field_validator("pdf_files", "image_files", "extra_files", mode="before")
    @classmethod
    def _coerce_json(cls, v: Any) -> Any:
        if v is None:
            return []
        if isinstance(v, str):
            return [a.model_dump() for a in _parse_attachments(v)]
        return v


class AClassAchievementListResponse(BaseModel):
    items: list[AClassAchievementRead]
    total: int
    research_categories: list[str]
    levels: list[str]
